Initial attempt to search within apps from KISS

losingkeys opened the pull request:
WARNING: this is probably not ready to be merged yet, because of the issues below. It's good enough for daily use though, and is probably worth discussing.

See https://github.com/Neamar/KISS/issues/157

Issues:
- Not all apps provide a search activity
  - The most notable missing apps are:
    - The Google app
    - The DuckDuckGo app
    - Browsers
    - Wikipedia
- Some apps provide a search activity, but it crashes
  - For example Google Newsstand
- Some apps provide a search activity, but it requires more than just a
  query
  - For example Gmail
- Some apps provide a search activity, but when you give it a query, it
  opens without preforming a search
  - For example Podcast Addict, Amazon Kindle, and Coursera
- Search results are not retained in the recent apps list
  - I don't care if the searches are stored, but it'd be nice if the
    apps that were searched showed up. This proved difficult to do
    because Providers all maintain their own list of recent Pojos

losingkeys commented:
To be clear, I still think this is a good idea, even with all the issues I ran into when trying to implement it. It's nice to be able to type a query without having to open the app, wait for it to load, and tapping the search bar, then being bothered by autocomplete as you type. It's also faster.

I'd like to find a way to search in web browsers, and maybe the ddg app, before merging this (if you end up wanting to merge it). Let me know what you think if you get a chance to try it out :)

Neamar commented:
Just tested:
![2015-10-12 09 55 31](https://cloud.githubusercontent.com/assets/536844/10425471/35bb0b60-70d8-11e5-8b89-b6f82c5197f4.png)

Works for everything but the calendars here.

I think we should open an issue with DDG so they add the missing searchactivity; this seems like the proper behavior and it's probably a nice thing to add to their app anyway.

Search results could be retained in history (actually i made something on #132 but reverted it since it was not so useful at the time)

Regarding the bugs... not sure what to do with them. I guess that's because other app behave improperly (thinking at you, Coursera!) dropping the initial query.

All in all, this is looking good already :) but as you said, going forward won't be easy from now on :\

losingkeys commented:
Another small issue I may not have mentioned: for the most part with the search activities we get the app names, but sometimes they just say "Search", or something similar. This will be an issue for users with icons turned off (I think I read this will be an option our already it's).

As to opening issues/pulls in other apps, I kind of figured this might happen. I'd like for firefox and Wikipedia to provide search activities. Also I thin Gmail's activity requires more than just a query. I don't want to special case things for Gmail, so maybe we can open a support ticket there.

Finally I was thinking it'd be nice if the history showed recent apps, not searches. Searches would just add the searched app to the list as if it was opened normally. What do you think? I think having recent searches lying around wouldn't look as good as just the apps, plus if you recently navigated away from an app you could just jump back in recently and see your results anyway.

Like I said before I like how this works... But it's a maintenance issue and probably a (small?) source of bugs. I wonder what we can do to mitigate that.

losingkeys commented:
One more thing I forgot: calendar search is working for me... maybe we have different versions of android?

Neamar commented:
Weird. I'm on Android 5.1.

Ok, let's open some PR on other apps!

Regarding the "Search" string issue, maybe we can go for something more generic? Like compare the string with our R.string.Search: if equal, revert to generic "Search on [app name]"?

losingkeys commented:
My goal with the search string was to be as short as possible, because the
query will take up a large part of the string. I think I'm getting the
activity name right now, but I bet I can get the app name. The only other
thing (with the string) is tranlations. I did my best to delete the
duplicates and retain the translations that quote things differently than
english, but it wouldn't hurt to look it over.
On Oct 13, 2015 09:00, "Matthieu Bacconnier" <notifications@github.com>
wrote:

> Weird. I'm on Android 5.1.
>
> Ok, let's open some PR on other apps!
>
> Regarding the "Search" string issue, maybe we can go for something more
> generic? Like compare the string with our R.string.Search: if equal, revert
> to generic "Search on [app name]"?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Neamar/KISS/pull/288#issuecomment-147723609>.
>

losingkeys commented:
Fixed in recent commits:
- No more 'Search ... "query"' in search results list
  - Using app names instead of the search activity name
- Rebased on Neamar/master to fix the build

Still TODO:
- Ensure all translations are acceptable
  - I tried to leave the ones that have special quoting (e.g. «like this»), but I'm sure some languages would prefer the app before the query, and might not use ':' in the same way english does.
  - We could remove all translations except for english and ask translators to re-submit if they don't like the quoting/use of the ':'. I think we at least need review. I'm not sure how to solicit this for all translated languages on weblate though.
- Fix the settings page
  - It still says "Enable Google Web Search"
  - Should I make a page where you can disable some providers? Or is it all or nothing?
- Documentation. We should put docs somewhere noting this new functionality, how to add providers, and to _please_ report bugs if your app isn't available or has problems.
  - I'd also like to add some gifs... but that's for another PR

Can you see if your calendar is working now? I haven't changed anything... but mine is. I have android 5.1 too (LG G4).

Finally, what do you think we should do about missing support for the Google App, DDG app, and Browsers? Merging now removes this functionality if/until we can get them to change their apps _or_ add special cases for these apps. Should we wait until we do one of those things to merge?

losingkeys commented:
Another todo: don't add disabled apps to the list.

Neamar commented:
* For translation: let's add a new string and remove the previous one, so we know which one should be done again.
* For now, let's enable all or nothing-- we can update the string to read "search in various apps" though
* Documentation: you're right. Maybe in the README? It's not perfect, but it's a good start.
* DDG / etc : let's open an issue on their side, and keep our code clean for now.
* Sidenote: can we also add Google Maps to the list? It'd be very useful for me :)
* Just switched to Android M. Calendars still not working, displaying a blank screen :(

![2015-10-20 14 14 44](https://cloud.githubusercontent.com/assets/536844/10609841/bf385cb2-7745-11e5-8d90-35987f1f1072.png)

Neamar commented:
(also, do we need contact search? Seems a bit redundant witjh Kiss built-in behavior, except for very specific use case)

Neamar commented:
(hum, thinking back on this -- we may need to add at least one special case for the default search [google one currently] as this is a feature that's used today)

losingkeys commented:
Updated a bit. Still TODO:
- Add documentation about adding and troubleshooting search providers
  - CONTRIBUTING would be a [good place to put this](https://github.com/blog/1184-contributing-guidelines). Mind adding some info for contributors? That way the README could contain only info relevant to users, and CONTRIBUTING could have more details that only potential contributors would be interested in
- Add support for Google/DDG apps
- Add support for browsers? Maybe wait on these.

Neamar commented:
I like the contributing idea :)

Also, I had some troubles with this branch as it is yet since i found myself unable to "search" from kiss on the internet -- it would open my contacts app when i was expecting Google or any other search engine.

losingkeys commented:
Sorry for the delay with this one. I'll probably be another few months, as I have another project I need to work on. From memory here's what I'd like to finish before this is merged:
- Rank searches by the app's popularity
- Show recently searched-in apps in the recents history
- Add support for Google/DDG apps and browsers

I'll let you know when I've made more progress.

Neamar commented:
Closing this PR for now (#429 is a good short term solution IMHO),

Will reopen this one when the Android ecosystem follows the Intent rules!
